EventFilter/DTRawToDigi/plugins/DTuROSRawToDigi.cc

```diff
@@ -0,0 +1,467 @@
+//
+//   Class: DTuROSRawToDigi
+//
+//   L1 DT uROS Raw-to-Digi
+//
+//
+//
+//   Author :
+//   C. Heidemann - RWTH Aachen
+//   J. Troconiz  - UAM
+//
+//
+//--------------------------------------------------
+
+#include "DTuROSRawToDigi.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "CondFormats/DTObjects/interface/DTReadOutMapping.h"
+#include "CondFormats/DataRecord/interface/DTReadOutMappingRcd.h"
+
+#include "DataFormats/DTDigi/interface/DTuROSControlData.h"
+#include "DataFormats/MuonDetId/interface/DTWireId.h"
+#include "EventFilter/DTRawToDigi/interface/DTROChainCoding.h"
+#include "EventFilter/Utilities/interface/DTCRC.h"
+
+#include <iostream>
+
+
+DTuROSRawToDigi::DTuROSRawToDigi(const edm::ParameterSet& pset) {
+
+  produces<DTDigiCollection>();
+//  produces<std::vector<DTuROSFEDData>>("DTuROSWords");
+  produces<std::vector<DTuROSFEDData>>();
+
+  DTuROSInputTag_ = pset.getParameter<edm::InputTag>("inputLabel");
+
+  debug_ = pset.getUntrackedParameter<bool>("debug", false);
+
+  feds_ = pset.getUntrackedParameter<std::vector<int> >("feds", std::vector<int>());
+ 
+  nfeds_ = feds_.size();
+
+  Raw_token = consumes<FEDRawDataCollection>(DTuROSInputTag_);
+
+}
+
+
+DTuROSRawToDigi::~DTuROSRawToDigi(){}
+
+
+void DTuROSRawToDigi::produce(edm::Event& e, const edm::EventSetup& c) {
+
+  DTDigiCollection digis;
+  std::vector<DTuROSFEDData> words;
+
+  if (!fillRawData(e, c, digis, words)) return;
+
+  auto uROSDTDigi_product = std::make_unique<DTDigiCollection>(digis);
+  auto uROSDTWord_product = std::make_unique<std::vector<DTuROSFEDData>>(words);
+
+  e.put(std::move(uROSDTDigi_product));
+  e.put(std::move(uROSDTWord_product));
+
+}
+
+
+bool DTuROSRawToDigi::fillRawData(edm::Event& e, const edm::EventSetup& c,
+                                  DTDigiCollection& digis, std::vector<DTuROSFEDData>& words) {
+
+  edm::Handle<FEDRawDataCollection> data;
+  e.getByToken(Raw_token, data);
+
+  edm::ESHandle<DTReadOutMapping> mapping;
+  c.get<DTReadOutMappingRcd>().get( mapping );
+
+  for (int w_i = 0; w_i < nfeds_; ++w_i) {
+    DTuROSFEDData fwords;
+    process(feds_[w_i], data, mapping, digis, fwords);
+    words.push_back(fwords);
+  }
+  
+  return true;
+}
+
+
+void DTuROSRawToDigi::process(int DTuROSFED,
+                              edm::Handle<FEDRawDataCollection> data,
+                              edm::ESHandle<DTReadOutMapping> mapping,
+                              DTDigiCollection& digis,
+                              DTuROSFEDData& fwords) {
+
+
+  // Container
+  std::vector<long> DTuROSWordContainer;
+
+
+  // Header constituents
+  int BOEevTy, DTuROSId;
+
+
+  // Trailer constituents
+  int chkEOE, evtLgth, CRC;
+
+
+  // Hit counter
+  std::map<uint32_t, int> hitOrder;
+
+
+  //--> Header - line 1
+
+
+  FEDRawData dturosdata = data->FEDData(DTuROSFED); 
+  if ( dturosdata.size() == 0 ) return;
+
+
+  lineFED=dturosdata.data(); 
+  long dataWord = 0;
```

Comment:
Normally for bit manipulation unsigned should be preferred, instead of signed integers.
It is true that this whole class, as well as the similar ones from which it had been derived, relies on signed integers: let stick on it, then...